kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local valid_type_names = {
+  SRV     = TYPE_SRV,
+  A       = TYPE_A,
+  AAAA    = TYPE_AAAA,
+  CNAME   = TYPE_CNAME,
+  LAST    = TYPE_LAST,
+}
+
+local typstrs = {
+  [TYPE_SRV]      = "SRV",
+  [TYPE_A]        = "A",
+  [TYPE_AAAA]     = "AAAA",
+  [TYPE_CNAME]    = "CNAME",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local hitstrs = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "hit_cb",
+  [4] = "hit_stale",
+}
+
+-- server replied error from the DNS protocol
+local NAME_ERROR_CODE    = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_EC      = 100
+local CACHE_ONLY_ESTR    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = { errcode = CACHE_ONLY_EC, errstr = CACHE_ONLY_ESTR }
+local EMPTY_RECORD_EC    = 101
+local EMPTY_RECORD_ESTR  = "empty record received"
+
+
+-- APIs
+
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+  if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+    _M[k] = v
+  end
+end
+_M.TYPE_LAST = TYPE_LAST
+
+
+local tries_mt = { __tostring = cjson.encode }
+
+
+local function stats_init(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_count(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+-- lookup or set TYPE_LAST (the DNS record type from the last successful query)
+local function insert_last_type(cache, name, qtype)
+  local key = "last:" .. name
+  if typstrs[qtype] and cache:get(key) ~= qtype then
+    cache:set(key, { ttl = 0 }, qtype)
+  end
+end
+
+
+local function get_last_type(cache, name)
+  return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+  local hosts, err = parse_hosts(path)
+  if not hosts then
+    log(WARN, "Invalid hosts file: ", err)
+    hosts = {}
+  end
+
+  if not hosts.localhost then
+    hosts.localhost = {
+      ipv4 = "127.0.0.1",
+      ipv6 = "[::1]",
+    }
+  end
+
+  local function insert_answer(name, qtype, address)
+    if not address then
+      return
+    end
+
+    local ttl = LONG_LASTING_TTL
+
+    local key = name .. ":" .. qtype
+    local answers = {
+      ttl = ttl,
+      expire = now() + ttl,
+      {
+        name = name,
+        type = qtype,
+        address = address,
+        class = 1,
+        ttl = ttl,
+      },
+    }
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(key, nil, function()
+      return answers, nil, ttl
+    end)
+  end
+
+  for name, address in pairs(hosts) do
+    name = name:lower()
+    if address.ipv4 then
+      insert_answer(name, TYPE_A, address.ipv4)
+      insert_last_type(cache, name, TYPE_A)
+    end
```

Comment:
```suggestion
    end

```

Could you fix other similar places as well?